packages/shell-api/src/decorators.ts

```diff
@@ -34,6 +50,55 @@ export class ShellApiClass implements ShellApiInterface {
   }
 }
 
+// For classes like Collection, it can be useful to attach information to the
+// result about the original data source, so that downstream consumers of the
+// shell can e.g. figure out how to edit a document returned from the shell.
+// To that end, we wrap the methods of a class, and report back how the
+// result was generated.
+// We also attach the `shellApiType` and `asShellResult` properties to the
+// return type (if that is possible and they are not already present), so that
+// we can also provide sensible information for methods that do not return
+// shell classes, like db.coll.findOne() which returns a Document (i.e. a plain
+// JavaScript object).
+function wrapWithAddSourceToResult(fn: Function, functionName: string): Function {
+  function addSource<T extends {}>(result: T, obj: any, args: any[]): T {
+    if (typeof result === 'object' && result !== null) {
+      const resultSourceInformation: ShellResultSourceInformation = {
+        dataSource: obj._dataSourceIdentifier(),
```

Comment:
we don't have a concept of `dataSource` i think the same is called namespace in our domain area, is probably better to keep the language consistent and use that name.

Reply:
`namespace` sounds good :+1: